---
title: "GenmStats_week4"
author: "Alisa"
date: "2/27/2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Week 4 Quiz

Gene set enrichment and eQTL (working with multiple forms of data)

##Question 2: Load the Bottomly data with the following code and perform a differential expression analysis using 𝚕𝚒𝚖𝚖𝚊 with only the strain variable as an outcome. How many genes are differentially expressed at the 5% FDR level using Benjamini-Hochberg correction? What is the gene identifier of the first gene differentially expressed at this level (just in order, not the smallest FDR) ? (hint: the 𝚏𝚎𝚊𝚝𝚞𝚛𝚎𝙽𝚊𝚖𝚎𝚜 function may be useful)

1. After loading data, use limma to make matrix of edata, pdata, in reference to strain.
2. 

```{r Question 2}
library(Biobase)
library(limma)
con =url("http://bowtie-bio.sourceforge.net/recount/ExpressionSets/bottomly_eset.RData")
load(file=con)
close(con)
bot = bottomly.eset
pdata_bot=pData(bot)
fdata_bot = featureData(bot)
edata_bot = exprs(bot)
fdata_bot = fdata_bot[rowMeans(edata_bot) > 5]
edata_bot = edata_bot[rowMeans(edata_bot) > 5, ]
edata_bot = log2(edata_bot+1)
mod = model.matrix(~ pdata_bot$strain)
fit_limma = lmFit(edata_bot,mod)
ebayes_limma = eBayes(fit_limma)
limma_all = topTable(ebayes_limma, number = dim(edata_bot)[1], adjust.method = "BH", sort = "none")
sum(limma_all$adj.P.Val < 0.05)
limma_pval_adj = limma_allt$adj.P.Val
head(limma_all, 1)
```

## Question 3: Use the 𝚗𝚞𝚕𝚕𝚙 and 𝚐𝚘𝚜𝚎𝚚 functions in the 𝚐𝚘𝚜𝚎𝚚 package to perform a gene ontology analysis. What is the top category that comes up as over represented? (hint: you will need to use the genome information on the genome from question 1 and the differential expression analysis from question 2.

1. Intention: identify enhanced gene category in bottomly data set using nullp and goseq functions.
2. Genome to use: mm9
3. de data is limma_all > extract genes with adj_Pvalue < 0.05
4. 

```{r pressure, echo=FALSE}
genes = as.integer(limma_pval_adj < 0.05)
not_na = !is.na(genes)
names(genes) = rownames(edata_bot)
genes = genes[not_na]
pwf <- nullp(genes,"mm9","ensGene")
GO.wall=goseq(pwf,"mm9","ensGene")
strain_only_top10 <- GO.wall[1:10,1]
```

## Question 5: How many of the top 10 overrepresented categories are the same for the adjusted and unadjusted analysis?

Extra info: Load the Bottomly data with the following code and perform a differential expression analysis using 𝚕𝚒𝚖𝚖𝚊 and treating strain as the outcome but adjusting for lane as a factor. Then find genes significant at the 5% FDR rate using the Benjamini Hochberg correction and perform the gene set analysis with 𝚐𝚘𝚜𝚎𝚚 following the protocol from the first 4 questions. 

```{r Question 5}
mod_adj = model.matrix(~ pdata_bot$strain + pdata_bot$lane.number)
fit_limma = lmFit(edata_bot,mod)
ebayes_limma = eBayes(fit_limma)
limma_all_adj = topTable(ebayes_limma, number = dim(edata_bot)[1], adjust.method = "BH", sort = "none")
limma_pval_adj1 = limma_all_adj$adj.P.Val
genes = as.integer(limma_pval_adj1 < 0.05)
not_na = !is.na(genes)
names(genes) = rownames(edata_bot)
genes = genes[not_na]
pwf <- nullp(genes,"mm9","ensGene")
GO.wall=goseq(pwf,"mm9","ensGene")
strain_lane_top10 <- GO.wall[1:10,1]
strain_lane_top10
strain_only_top10
